riscv/triggers.cc

```diff
@@ -525,6 +552,24 @@ std::optional<match_result_t> module_t::detect_memory_access_match(operation_t o
   return std::nullopt;
 }
 
+std::optional<match_result_t> module_t::detect_icount_match() noexcept
+{
+  for (auto trigger: triggers)
+    trigger->stash_read_values();
+
+  state_t * const state = proc->get_state();
+  if (state->debug_mode)
+    return std::nullopt;
+
+  std::optional<match_result_t> ret = std::nullopt;
+  for (auto trigger: triggers) {
+    auto result = trigger->detect_icount_match(proc);
+    if (result.has_value())
+      ret = result;
+  }
+  return ret;
```

Comment:
There is one other consideration that I just now realized. In the debug spec 5.3 "Priority" it says about multiple triggers firing simultaneously:

> If one of these triggers has the “enter Debug Mode” action (1) and another trigger has the “raise a breakpoint exception” action (0), the preferred behavior is to have both actions take place. [...] If this is not implemented, then the hart must enter Debug Mode and ignore the breakpoint exception.

The above loop appears to return the action from the highest-index breakpoint that triggers, regardless of others that might have a different action.

It seems that `module_t::detect_memory_access_match` is similarly broken. I'll try to fix that in a separate PR.

Reply:
I think I misunderstood. I see you added commit fb8a8d5cbc6665e9dd89961dab2cf24c33d1ed80 to address this in this PR. I also opened #1231 because I thought you were saying you'd fix this in a future PR.

I don't have good test coverage of this case, but AFAICT this will fix #1231.